enterprise/backend/src/metabase_enterprise/workspaces/dependencies.clj

```diff
@@ -0,0 +1,308 @@
+                        (into (lib/all-source-table-ids query))
+                        (into (lib/all-implicitly-joined-table-ids query)))
+        table-refs  (batch-table-refs-from-ids table-ids)]
+    (vec (keep table-refs table-ids))))
+
+(defn- inputs-from-native-query
+  "Extract table refs from a native query. Uses driver/native-query-table-refs to get
+   schema+table directly, then batch looks up table_ids for tables that exist.
+   Uses the same default schema logic as find-table-or-transform."
+  [query database-id]
+  (let [driver         (:engine (lib.metadata/database query))
+        parsed-refs    (driver/native-query-table-refs driver query)
+        ;; Batch lookup existing table IDs (handles default schema logic)
+        table-id-map   (batch-lookup-table-ids driver database-id parsed-refs)]
+    (vec (for [{:keys [schema table]} parsed-refs
+               :let [table-id (get table-id-map [database-id schema table])]]
+           (cond-> {:database_id database-id
+                    :schema      schema
+                    :table       table}
+             table-id (assoc :table_id table-id))))))
+
+(defn- inputs-from-query
+  "Extract table refs from a query (MBQL or native)."
+  [query database-id]
+  (if (lib/native-only-query? query)
+    (inputs-from-native-query query database-id)
+    (inputs-from-mbql-query query)))
+
+(defn- inputs-from-python-transform
+  "Extract table refs from a python transform's source-tables.
+   Python transforms require tables to exist (they map name -> table_id). Batch lookup."
+  [source-tables]
+  (let [table-ids  (set (vals source-tables))
+        table-refs (batch-table-refs-from-ids table-ids)]
+    (vec (keep table-refs (vals source-tables)))))
+
+(mu/defn analyze-entity :- ::analysis
+  "Analyze a workspace entity to find its dependencies.
+
+   Arguments:
+   - entity-type: keyword, must be :transform (asserted)
+   - entity: the workspace transform map with :source and :target
+
+   Returns:
+   {:output {:database_id int :schema string :table string}
+    :inputs [{:database_id int :schema string :table string} ...]}"
+  [entity-type :- :keyword
+   entity      :- :map]
+  (ws.u/assert-transform! entity-type)
+  (let [{{:keys [query source-tables], source-type :type} :source
+         target :target} entity
+        database-id (:database target)
+        inputs (cond
+                 (= (keyword source-type) :query)
+                 (inputs-from-query query database-id)
+
+                 (= (keyword source-type) :python)
+                 (inputs-from-python-transform source-tables)
+
+                 :else
+                 (do (log/warnf "Don't know how to analyze deps of transform with source type '%s'" source-type)
+                     []))]
+    {:output (output-from-target target)
+     :inputs inputs}))
+
+;;; ---------------------------------------- Write helpers ----------------------------------------
+
+(defn- upsert-workspace-output!
```

Comment:
Problem: this function has race conditions.

If there is no constraint in the db, it could create duplicates.
If there is a constraint, it could fail on conflict without retrying.

This is why we have `app-db/update-or-insert!`